proposals/168/making_spark_visitors_extensible.md

```diff
@@ -0,0 +1,271 @@
+an `OpenLineageContext` class, defined as
+
+```java
+@Value
+public class OpenLineageContext {
+    Collection<QueryPlanVisitor<LogicalPlan>> inputDatasetQueryPlanVisitors;
+    Collection<QueryPlanVisitor<LogicalPlan>> outputDatasetQueryPlanVisitors;
+    Collection<CustomFacetBuilder<Object, OpenLineage.InputDatasetFacet>> inputDatasetFacetBuilders;
+    Collection<CustomFacetBuilder<Object, OpenLineage.OutputDatasetFacet>> outputDatasetFacetBuilders;
+    Collection<CustomFacetBuilder<Object, OpenLineage.DatasetFacet>> datasetFacetBuilders;
+    Collection<CustomFacetBuilder<Object, OpenLineage.RunFacet>> runFacetBuilders;
+    Collection<CustomFacetBuilder<Object, OpenLineage.JobFacet>> jobFacetBuilders;
+    SparkContext sparkContext;
+}
+```
+
+The `OpenLineageEventListenerFactory#setOpenLineageContext` method will be called once the concrete service 
+implementation has been constructed, before the first `createX` method is called.
+
+Each of the `createX` methods will be called exactly once during a Spark application. There's no guarantee regarding the
+order in which the `create` methods will be called. Thus, the constructed builders should make no assumptions about the
+state of the `OpenLineageContext` when the `create` method is called. E.g., the `inputDatasetFacetBuilders` should not 
+assume that all of the `inputDatasetQueryPlanVisitors` have been instantiated or that the `Collection` returned by the
+getter has been fully populated. Builders that need to depend on the ability to delegate to other builders should store
+a reference to the `OpenLineageContext` and call the appropriate getter at the time the delegate is actually needed.
+
+The builders returned are all some form of `PartialFunction` - they must all define an `isDefinedAt(T event)` method
+which determines the type of argument the builder can respond to. The `createInputDatasetBuilder` and 
+`createOutputDatasetBuilder` methods return `PartialFunction`s that construct either an `InputDatasetBuilder` or an
+`OutputDatasetBuilder`, respectively. Builders are returned so that facets constructed by the other `PartialFunction`s
+can be attached to the dataset before the immutable `InputDataset` or `OutputDataset` is returned. It is expected that
+the `CustomFacet`s returned are already immutable.
+
+For each `SparkListener` event, an `OpenLineage.RunEvent` will be constructed by 
+1. Construct `RunFacet`s by invoking each of the `runFacetBuilders`
+2. Construct `JobFacet`s by invoking each of the `jobFacetBuilders`
+3. Construct `InputDatasetBuilder`s by invoking each of the `inputDatasetBuilder`s.
+4. _If_ any `InputDatasetBuilder`s are returned, invoke the `inputDatasetFacetBuilder`s and attach any returned facets to
+   the `InputDatasetBuilder`s
+5. _If_ any `InputDatasetBuilder`s are returned, invoke the `datasetFacetBuilder`s and attach any returned facets to 
+    the `InputDatasetBuilder`s
```

Comment:
> where will facets created by datasetFacetBuilders be attached? To both of them?

This is a good question. From my experimenting, a `Stage` that reads the raw data will only contain one input dataset and no output datasets. A job will have only one output dataset.

A possible exception to this includes dynamic datasets, such as dynamic partitions in Hive or dynamic topics in Kafka. E.g., a record can contain the name of the Kafka topic as a column and the Kafka topic _should_ be counted as a distinct dataset. However, we currently don't support this behavior and I think adding support for this would not allow us to rely on facets created by inspecting events on the driver. E.g., the `recordsWritten` field in the `TaskMetrics` would be the sum total of all records written to Kafka, not those written to a specific topic. To collect the actual number written to a specific topic would involve counting those records on the executor. This is work that would be very custom - it is supported by this framework, but the implementation work would likely be done by an application owner, not us.